feature: keyboard accessibility for tabbed UIs (fixes #526)

greatislander opened the pull request:
Hi @mouse-reeve! Took me longer than I'd hoped to get to this but here it is. I've used a slightly modified version of https://www.npmjs.com/package/@zachleat/seven-minute-tabs, which wraps tab markup in a web component which registers all the necessary event handlers. If that's not ideal (it doesn't support IE) I can rework it to use more vanilla JavaScript— let me know your preference.

rkingett commented:
@leonstafford check this out, but, personally, more vanilla JavaScript would be the best, as, even today, there are still people using IE. Although, I am not keeping up with traffic and such. Let's let Mouse decide.

mouse-reeve commented:
Thank you for getting this going! It's really nice to be able to move through the tabs with arrow keys. I would prefer to keep IE compatibility as much as possible, so a vanilla js approach is more appealing to me. I don't log browser usage so I don't know if anyone is using IE, though. 

My little quibble is that, while moving the keyboard focus to the panel when a tab is selected is great, I find the visual scroll-down jump to the panel jarring, and would prefer the scroll position to not change.

greatislander commented:
> Thank you for getting this going! It's really nice to be able to move through the tabs with arrow keys. I would prefer to keep IE compatibility as much as possible, so a vanilla js approach is more appealing to me. I don't log browser usage so I don't know if anyone is using IE, though.
> 
> My little quibble is that, while moving the keyboard focus to the panel when a tab is selected is great, I find the visual scroll-down jump to the panel jarring, and would prefer the scroll position to not change.

Hi @mouse-reeve, I'll rework it to use a more backwards-compatible vanilla JS implementation and see what I can do about the scroll jump. Thanks!

mouse-reeve commented on bookwyrm/templates/feed/feed_layout.html:
```diff
@@ -10,56 +10,58 @@ <h2 class="title is-5">{% trans "Your books" %}</h2>
-                    <p>
-                        {{ shelf.name }}
+            {% for book in shelf.books %}
+            <div class="suggested-tabs card" role="tabpanel" id="book-{{ book.id }}"{% if shelf_counter != 1 or not forloop.first %} hidden{% endif %}>
```
Am I missing something or has accesslint gotten confused about what elements are what?

greatislander replied:
I think this is a valid issue, @mouse-reeve — the tab panel should have an aria-labelledby attribute which references the controlling tab's title text. However, in this case that's complicated by the fact that the tab is an image— I'm going to try to improve this in my ongoing work.

greatislander commented on bookwyrm/templates/snippets/create_status.html:
```diff
@@ -2,34 +2,36 @@
-    {% include 'snippets/create_status_form.html' with type='review' %}
-    {% endwith %}
-</div>
+    <div class="tab-option-{{ book.id }}" id="review-{{ book.id }}" role="tabpanel" tabindex="0" aria-labelledby="tab-review-{{ book.id }}">
```
As for this I think it is not correctly identifying the controlling tab by ID. I'm not sure why— I'll do some testing with voiceover and see what happens.

greatislander replied:
This doesn't appear to be a problem in screen reader testing, I'm going to mark it as resolved.

greatislander commented:
Okay, @mouse-reeve, this is updated with the following changes:

- Rewrote tab handler as vanilla Javascript, integrated into `shared.js`.
- Fixed focus jump scroll into tab panel contents.
- Added `aria-label` to book tabs.

As for the last point, this touches on the [1book_cover.html`](https://github.com/mouse-reeve/bookwyrm/blob/263b13f1576a901a995a414aa45f97ae2f2917b1/bookwyrm/templates/snippets/book_cover.html) snippet.

I'll try to clarify my thinking here and see if this makes sense.

In contexts where the book cover snippet is not being used as a tab control, having the accessible name be the image alt text makes sense. However, overriding the accessible name to just the book title is preferable in this context because the tab should be named "Book Title" rather than "Book title cover (hardcover, 2020)". For an example of this, compare the four tabs demonstratedd here: https://codepen.io/greatislander/pen/oNYqExm

Using an `aria-label` attribute and the `<a role="tab">` element allows the tab to have a proper accessible name while letting the book cover snippet fall back to its alt text in other contexts.

(One other note is that I've removed the title attribute on the book image in the book cover snippet as it's redundant alongside the alt text.)

Let me know if this makes sense! Happy to make further adjustments as needed.

rkingett commented:
It appears to be fine on a windows side, too!

mouse-reeve commented:
this is so slick! it's a pleasure to use, thank you!
